sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/PaginationCarbonReader.java

```diff
@@ -0,0 +1,296 @@
+import org.apache.carbondata.core.constants.CarbonCommonConstants;
+import org.apache.carbondata.core.indexstore.BlockletDetailInfo;
+import org.apache.carbondata.hadoop.CarbonInputSplit;
+import org.apache.carbondata.sdk.file.cache.BlockletRows;
+
+import org.apache.hadoop.mapreduce.InputSplit;
+
+/**
+ * CarbonData SDK reader with pagination support
+ */
+@InterfaceAudience.User
+@InterfaceStability.Evolving
+public class PaginationCarbonReader<T> extends CarbonReader<T> {
+  // Splits based the file present in the reader path when the reader is built.
+  private List<InputSplit> allBlockletSplits;
+
+  // Rows till the current splits stored as list.
+  private List<Long> rowCountInSplits;
+
+  // Reader builder used to create the pagination reader, used for building split level readers.
+  private CarbonReaderBuilder readerBuilder;
+
+  private boolean isClosed;
+
+  // to store the rows of each blocklet in memory based LRU cache.
+  // key: unique blocklet id
+  // value: BlockletRows
+  private CarbonLRUCache cache =
+      new CarbonLRUCache(CarbonCommonConstants.CARBON_MAX_PAGINATION_LRU_CACHE_SIZE_IN_MB,
+          CarbonCommonConstants.CARBON_MAX_PAGINATION_LRU_CACHE_SIZE_IN_MB_DEFAULT);
+
+  /**
+   * Call {@link #builder(String)} to construct an instance
+   */
+
+  PaginationCarbonReader(List<InputSplit> splits, CarbonReaderBuilder readerBuilder) {
+    // Initialize super class with no readers.
+    // Based on the splits identified for pagination query, readers will be built for the query.
+    super(null);
+    this.allBlockletSplits = splits;
+    this.readerBuilder = readerBuilder;
+    // prepare the mapping.
+    rowCountInSplits = new ArrayList<>(splits.size());
+    long sum = ((CarbonInputSplit) splits.get(0)).getDetailInfo().getRowCount();
+    rowCountInSplits.add(sum);
+    for (int i = 1; i < splits.size(); i++) {
+      // prepare a summation array of row counts in each blocklet,
+      // this is used for pruning with pagination vales.
+      // At current index, it contains sum of rows of all the blocklet from previous + current.
+      sum += ((CarbonInputSplit) splits.get(i)).getDetailInfo().getRowCount();
+      rowCountInSplits.add(sum);
+    }
+  }
+
+  /**
+   * Pagination query with from and to range.
+   *
+   * @param from must be greater than 0 and <= to
+   * @param to must be >= from and not outside the total rows
+   * @return array of rows between from and to (inclusive)
+   * @throws Exception
+   */
+  public Object[] read(long from, long to) throws IOException, InterruptedException {
+    if (isClosed) {
+      throw new RuntimeException("Pagination Reader is closed. please build again");
+    }
+    if (from < 1) {
+      throw new IllegalArgumentException("from row id:" + from + " is less than 1");
+    }
+    if (from > to) {
+      throw new IllegalArgumentException(
+          "from row id:" + from + " is greater than to row id:" + to);
+    }
+    if (to > getTotalRows()) {
```

Comment:
Is parameter starts from 0 or 1? should mention in line 85

Reply:
In line 86, I have mentioned 'to' must be >= 'from'
so, this has to be greater than 0.

If still need to mention. I can add.

Reply:
added